File: check-all-pages.js
import puppeteer from 'puppeteer';
import fs from 'fs/promises';

// 26개 전체 페이지 목록
const pages = [
  // 홈
  { name: '홈페이지', url: 'https://doha.kr/' },
  
  // 심리테스트 (6개)
  { name: '심리테스트 메인', url: 'https://doha.kr/tests/' },
  { name: 'MBTI 소개', url: 'https://doha.kr/tests/mbti/' },
  { name: 'MBTI 테스트', url: 'https://doha.kr/tests/mbti/test.html' },
  { name: 'MBTI 결과', url: 'https://doha.kr/tests/mbti/result.html' },
  { name: 'Teto-Egen 소개', url: 'https://doha.kr/tests/teto-egen/' },
  { name: 'Teto-Egen 테스트', url: 'https://doha.kr/tests/teto-egen/test.html' },
  { name: 'Teto-Egen 결과', url: 'https://doha.kr/tests/teto-egen/result.html' },
  { name: 'Love DNA 소개', url: 'https://doha.kr/tests/love-dna/' },
  { name: 'Love DNA 테스트', url: 'https://doha.kr/tests/love-dna/test.html' },
  { name: 'Love DNA 결과', url: 'https://doha.kr/tests/love-dna/result.html' },
  
  // 운세 (6개)
  { name: '운세 메인', url: 'https://doha.kr/fortune/' },
  { name: '오늘의 운세', url: 'https://doha.kr/fortune/daily/' },
  { name: 'AI 사주팔자', url: 'https://doha.kr/fortune/saju/' },
  { name: 'AI 타로 리딩', url: 'https://doha.kr/fortune/tarot/' },
  { name: '별자리 운세', url: 'https://doha.kr/fortune/zodiac/' },
  { name: '띠별 운세', url: 'https://doha.kr/fortune/zodiac-animal/' },
  
  // 실용도구 (4개)
  { name: '실용도구 메인', url: 'https://doha.kr/tools/' },
  { name: 'BMI 계산기', url: 'https://doha.kr/tools/bmi/' },
  { name: '글자수 세기', url: 'https://doha.kr/tools/text-counter.html' },
  { name: '연봉계산기', url: 'https://doha.kr/tools/salary-calculator.html' },
  
  // 기타 (5개)
  { name: '소개 페이지', url: 'https://doha.kr/about.html' },
  { name: '문의하기', url: 'https://doha.kr/contact.html' },
  { name: 'FAQ', url: 'https://doha.kr/faq.html' },
  { name: '개인정보처리방침', url: 'https://doha.kr/privacy.html' },
  { name: '이용약관', url: 'https://doha.kr/terms.html' }
];

async function checkAllPages() {
  console.log('🔍 doha.kr 26개 페이지 완전 검사 시작...\n');
  
  const browser = await puppeteer.launch({ 
    headless: 'new',
    args: ['--no-sandbox'] 
  });
  
  const results = [];
  
  for (let i = 0; i < pages.length; i++) {
    const pageInfo = pages[i];
    console.log(`[${i+1}/26] ${pageInfo.name} 검사 중...`);
    
    const page = await browser.newPage();
    const errors = [];
    const warnings = [];
    
    // 콘솔 메시지 수집
    page.on('console', msg => {
      if (msg.type() === 'error') {
        errors.push(msg.text());
      } else if (msg.type() === 'warning') {
        warnings.push(msg.text());
      }
    });
    
    // 페이지 에러 수집
    page.on('pageerror', error => {
      errors.push(error.message);
    });
    
    try {
      await page.goto(pageInfo.url, {
        waitUntil: 'networkidle2',
        timeout: 30000
      });
      
      // 2초 대기
      await new Promise(r => setTimeout(r, 2000));
      
      // 페이지 상태 체크
      const pageState = await page.evaluate(() => {
        return {
          title: document.title,
          hasContent: document.body && document.body.innerText.length > 100,
          hasCSS: document.styleSheets.length > 0,
          hasJS: document.scripts.length > 0,
          fontFamily: window.getComputedStyle(document.body).fontFamily,
          isPretendard: window.getComputedStyle(document.body).fontFamily.includes('Pretendard'),
          hasKorean: /[가-힣]/.test(document.body.innerText),
          buttonsWork: document.querySelectorAll('button, .btn').length,
          formsWork: document.querySelectorAll('form').length,
          linksWork: document.querySelectorAll('a[href]').length,
          imagesLoaded: Array.from(document.images).filter(img => img.complete).length,
          imagesBroken: Array.from(document.images).filter(img => !img.complete).length
        };
      });
      
      // 상태 판정
      let status = '✅';
      let statusText = '정상';
      
      if (errors.length > 5) {
        status = '❌';
        statusText = '심각';
      } else if (errors.length > 0) {
        status = '⚠️';
        statusText = '경고';
      } else if (!pageState.hasContent) {
        status = '❌';
        statusText = '빈페이지';
      } else if (!pageState.isPretendard) {
        status = '⚠️';
        statusText = '폰트미적용';
      }
      
      console.log(`   ${status} ${statusText} - 오류: ${errors.length}개`);
      
      results.push({
        name: pageInfo.name,
        url: pageInfo.url,
        status: statusText,
        errors: errors.length,
        warnings: warnings.length,
        state: pageState
      });
      
    } catch (error) {
      console.log(`   ❌ 로드 실패: ${error.message}`);
      results.push({
        name: pageInfo.name,
        url: pageInfo.url,
        status: '실패',
        errors: 999,
        warnings: 0,
        error: error.message
      });
    }
    
    await page.close();
  }
  
  // 결과 저장
  await fs.writeFile(
    'all-pages-check-result.json',
    JSON.stringify(results, null, 2)
  );
  
  // 요약 출력
  console.log('\n' + '='.repeat(60));
  console.log('📊 26개 페이지 검사 완료');
  console.log('='.repeat(60));
  
  const normal = results.filter(r => r.status === '정상').length;
  const warning = results.filter(r => r.status === '경고' || r.status === '폰트미적용').length;
  const critical = results.filter(r => r.status === '심각' || r.status === '실패' || r.status === '빈페이지').length;
  
  console.log(`✅ 정상: ${normal}개`);
  console.log(`⚠️ 경고: ${warning}개`);
  console.log(`❌ 심각: ${critical}개`);
  
  // 문제 페이지 목록
  if (critical > 0) {
    console.log('\n❌ 심각한 문제가 있는 페이지:');
    results.filter(r => r.status === '심각' || r.status === '실패' || r.status === '빈페이지')
      .forEach(r => console.log(`   - ${r.name}: ${r.errors}개 오류`));
  }
  
  if (warning > 0) {
    console.log('\n⚠️ 경고가 있는 페이지:');
    results.filter(r => r.status === '경고' || r.status === '폰트미적용')
      .forEach(r => console.log(`   - ${r.name}: ${r.errors}개 오류`));
  }
  
  await browser.close();
  console.log('\n✅ 검사 완료. 결과: all-pages-check-result.json');
}

checkAllPages().catch(console.error);

File: check_all_pages.py
import asyncio
from playwright.async_api import async_playwright
import xml.etree.ElementTree as ET
import json
from datetime import datetime
import os
import sys

# Set UTF-8 encoding for Windows console
if sys.platform == "win32":
    sys.stdout.reconfigure(encoding='utf-8')

async def check_page(page, url):
    """Check a single page for errors"""
    results = {
        "url": url,
        "status": "OK",
        "errors": [],
        "warnings": [],
        "console_errors": [],
        "network_errors": [],
        "missing_resources": []
    }
    
    # Collect console messages
    console_logs = []
    page.on("console", lambda msg: console_logs.append({
        "type": msg.type, 
        "text": msg.text
    }))
    
    # Collect network errors
    failed_requests = []
    page.on("requestfailed", lambda request: failed_requests.append({
        "url": request.url,
        "failure": request.failure
    }))
    
    try:
        # Navigate to page
        response = await page.goto(url, wait_until='networkidle', timeout=30000)
        
        # Check HTTP status
        if response.status >= 400:
            results["status"] = "ERROR"
            results["errors"].append(f"HTTP {response.status} error")
        
        # Wait for content to load
        await page.wait_for_timeout(2000)
        
        # Check for console errors
        for log in console_logs:
            if log["type"] in ["error", "warning"]:
                results["console_errors"].append(log)
                if log["type"] == "error":
                    results["status"] = "ERROR"
        
        # Check for network errors
        for failed in failed_requests:
            results["network_errors"].append(failed)
            results["status"] = "ERROR"
        
        # Check for missing CSS/JS files
        missing_resources = await page.evaluate('''() => {
            const errors = [];
            
            // Check CSS files
            const links = document.querySelectorAll('link[rel="stylesheet"]');
            links.forEach(link => {
                if (!link.sheet) {
                    errors.push({
                        type: 'CSS',
                        url: link.href,
                        message: 'Failed to load stylesheet'
                    });
                }
            });
            
            // Check for inline errors
            const errorElements = document.querySelectorAll('.error, .error-message');
            errorElements.forEach(el => {
                errors.push({
                    type: 'DOM',
                    message: el.textContent.trim()
                });
            });
            
            return errors;
        }''')
        
        if missing_resources:
            results["missing_resources"] = missing_resources
            results["status"] = "ERROR"
        
        # Check for common issues
        common_issues = await page.evaluate('''() => {
            const issues = [];
            
            // Check for broken images
            const images = document.querySelectorAll('img');
            images.forEach(img => {
                if (!img.complete || img.naturalWidth === 0) {
                    issues.push({
                        type: 'broken_image',
                        src: img.src
                    });
                }
            });
            
            // Check for empty content areas
            const mainContent = document.querySelector('main, .content, #content');
            if (mainContent && mainContent.textContent.trim().length < 50) {
                issues.push({
                    type: 'empty_content',
                    message: 'Main content area appears empty'
                });
            }
            
            return issues;
        }''')
        
        if common_issues:
            results["warnings"].extend(common_issues)
            
    except Exception as e:
        results["status"] = "ERROR"
        results["errors"].append(f"Page load error: {str(e)}")
    
    return results

async def check_all_pages():
    """Check all pages from sitemap"""
    # Parse sitemap
    tree = ET.parse('sitemap.xml')
    root = tree.getroot()
    namespace = {'sitemap': 'http://www.sitemaps.org/schemas/sitemap/0.9'}
    
    urls = []
    for url_element in root.findall('sitemap:url', namespace):
        loc = url_element.find('sitemap:loc', namespace)
        if loc is not None:
            urls.append(loc.text)
    
    print(f"Found {len(urls)} pages to check\n")
    
    # Check pages
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        context = await browser.new_context(
            viewport={'width': 1920, 'height': 1080},
            user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        )
        
        all_results = []
        error_count = 0
        
        for i, url in enumerate(urls, 1):
            print(f"[{i}/{len(urls)}] Checking: {url}")
            page = await context.new_page()
            
            try:
                result = await check_page(page, url)
                all_results.append(result)
                
                if result["status"] == "ERROR":
                    error_count += 1
                    print(f"  ❌ ERROR: {', '.join(result['errors'][:2])}")
                    if result["console_errors"]:
                        print(f"     Console errors: {len(result['console_errors'])}")
                    if result["network_errors"]:
                        print(f"     Network errors: {len(result['network_errors'])}")
                else:
                    print(f"  ✅ OK")
                    
            except Exception as e:
                print(f"  ❌ CRITICAL ERROR: {str(e)}")
                all_results.append({
                    "url": url,
                    "status": "CRITICAL",
                    "errors": [str(e)]
                })
                error_count += 1
            
            await page.close()
        
        await browser.close()
    
    # Generate report
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    report_path = f"page_check_report_{timestamp}.json"
    
    with open(report_path, 'w', encoding='utf-8') as f:
        json.dump({
            "timestamp": timestamp,
            "total_pages": len(urls),
            "error_count": error_count,
            "success_count": len(urls) - error_count,
            "results": all_results
        }, f, indent=2, ensure_ascii=False)
    
    print(f"\n{'='*60}")
    print(f"Check completed!")
    print(f"Total pages: {len(urls)}")
    print(f"Errors found: {error_count}")
    print(f"Success: {len(urls) - error_count}")
    print(f"Report saved to: {report_path}")
    
    # Show error summary
    if error_count > 0:
        print(f"\n{'='*60}")
        print("ERROR SUMMARY:")
        for result in all_results:
            if result["status"] == "ERROR":
                print(f"\n{result['url']}:")
                for error in result["errors"][:3]:
                    print(f"  - {error}")
                if result["console_errors"]:
                    print(f"  - {len(result['console_errors'])} console errors")
                if result["missing_resources"]:
                    print(f"  - {len(result['missing_resources'])} missing resources")

if __name__ == "__main__":
    asyncio.run(check_all_pages())